packages/eslint-plugin/src/rules/thenable-in-promise-aggregators.ts

```diff
@@ -0,0 +1,176 @@
+import {
+  isPromiseConstructorLike,
+  isPromiseLike,
+  isTypeAnyType,
+  isTypeUnknownType,
+} from '@typescript-eslint/type-utils';
+import type { TSESTree } from '@typescript-eslint/utils';
+import { AST_NODE_TYPES } from '@typescript-eslint/utils';
+import * as tsutils from 'ts-api-utils';
+import * as ts from 'typescript';
+
+import { createRule, getParserServices } from '../util';
+
+export default createRule({
+  name: 'thenable-in-promise-aggregators',
+  meta: {
+    docs: {
+      description:
+        'Disallow passing non-Thenable values to promise aggregators',
+      recommended: 'strict',
+      requiresTypeChecking: true,
+    },
+    messages: {
+      inArray:
```

Comment:
This is the main case I'm specifically looking to lint, and is the most common case where mistakes are made. There is no advantage to doing this from the perspective of control flow (i.e., how long the returned promise takes to fulfill), however, it does change the return value. However, whenever you really do want to provide a non-Thenable to change the control flow, you can always wrap it in `Promise.resolve` to get exactly the same behavior while making it way more clear how the aggregator will behave. This is still true even when considering if the returned promise is fulfilled synchronously vs asynchronously: these promise aggregators only fulfill synchronously when passed an empty array.

I would argue especially for all of these promise aggregators, the most important thing is making the control flow clear. Any non-Thenable value passed to a promise aggregator will be treated as `Promise.resolve(value)` according to the spec; see e.g., [PerformPromiseAll](https://tc39.es/ecma262/multipage/control-abstraction-objects.html#sec-performpromiseall). Thus given that a value in the list is known to be non-Thenable, there is a static conversion available that will make the control flow easier to understand, or you can be explicit and wrap the value in Promise.resolve, which jumps out at the reader and thus also makes the control flow easier to understand.

Given:

```ts
declare const aPromise: Promise<number>;
declare const bPromise: Promise<number;
declare const cNumber: number;
```

The following are all equivalent:


```ts
// all

Promise.all([aPromise, bPromise, cNumber])
// equivalent to
Promise.all([aPromise, bPromise, Promise.resolve(bNumber)])
// which resolves in the same amount of time as
Promise.all([aPromise, bPromise])

// previously working, now causes lint error
const [a, b, c] = await Promise.all([aPromise, bNumber]);

// suggested change
const [a, b] = await Promise.all([aPromise, bPromise]);
const c = cNumber;
```

```ts
// allSettled

Promise.allSettled([aPromise, bPromise, cNumber]);
// equivalent to
Promise.allSettled([aPromise, bPromise, Promise.resolve(cNumber)]);
// resolves in the same time as
Promise.allSettled([aPromise, bPromise]);

// previously working, now causes lint error
const [aPromise2, bPromise2, cPromise] = await Promise.allSettled([aPromise, bPromise, cNumber]);

// suggested change
const [aPromise2, bPromise2] = await Promise.allSettled([aPromise, bPromise]);
const cPromise = Promise.resolve(cNumber);
```

```ts
// race

await Promise.race([aPromise, bPromise, cNumber]);
// equivalent to
await Promise.race([aPromise, bPromise, Promise.resolve(cNumber)]);
// always resolves immediately (but never synchronously, as Promise.race is never synchronous)

// previously working, now causes lint error
// this is a if a is immediately resolved, b if b is immediately resolved, otherwise c
const aOrBOrC = await Promise.race([aPromise, bPromise, cNumber]);

// suggested change if you really did want the above behavior (most of the time you do NOT)
const aOrBOrC = await Promise.race([aPromise, bPromise, Promise.resolve(cNumber)]);

// suggested change if you actually intended to wait for something to happen
const aOrB = await Promise.race([aPromise, bPromise])

// note that most of the time this comes up, this is whats actually happening:
declare const aPromise: Promise<number>;
declare const bPromise: Promise<number>;
declare const cWrappedPromise: { promise: Promise<number> };

// lint error that catches a real subtle bug
const aOrBOrC = await Promise.race([aPromise, bPromise, cWrappedPromise]);

// correct code
const aOrBOrC = await Promise.race([aPromise, bPromise, cWrappedPromise.promise]);
```

```ts
// any is the same as race for the above purposes, the difference is how rejected values are handled
```

Reply:
I realized a situation where this option would be helpful! [<a href="https://deploy-preview-8094--typescript-eslint.netlify.app/play/#ts=5.3.3&fileType=.ts&code=CYUwxgNghgTiAEAzArgOzAFwJYHtVJAzAAsAKZGCALngGcMYtUBzAShoAUYcBbLWkAB4cAIwBW4DAD4A3ACg5KdNjzwQADyg8ADhBClW8AN5z48OBgr4uvfiAB0UCBFIBtU2YJEyAIgD0PqwANB5mPkzwGMQIfMDAej4hnl4kpP6BHgC6rPIAvkA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6fAC0SYEMAjJLWa1i0APYBbSskS0eAc3nRE8nvjHRUGSImjjokcGAC%2BIY0A&tsconfig=&tokens=false" title="Playground link with example function returning a Promise.all tuple">playground link</a>]

```ts
declare function fetch(url: string): Promise<object>;

function example() {
  return Promise.all([
    fetch("/"),
    "in the middle",
    fetch("/")
  ]);
}
```

If I was the developer here using `Promise.all` directly to avoid having to make & manage a `[object, string, object]` tuple, I'd be annoyed at the rule asking me to explicitly `Promise.resolve("in the middle")`. There's no benefit in this situation beyond consistency - at the cost of simplicity. 

But, I get what you're saying that this is really the main point of the rule. I think it's fine for now to leave as-is. Since it's only enabled in the strict config, it should just show up for users who have opted into the more strict opinions.

Reply:
Please do add an option that allows mixed arrays. I often parallel a bunch of tasks and sometimes tasks shift between being sync and async (for example, because it no longer dynamically imports a library, or it no longer reads a config file). I really don't want to break the symmetry in this case.

Reply:
In that case, +1ing Josh's request 🙂